packages/client-react-streaming/src/transportedQueryRef.ts

```diff
@@ -0,0 +1,126 @@
+> extends TransportedQueryRef<TData, TVariables> {
+  __transportedQueryRef: true | QueryRef<any, any>;
+  options: TransportedQueryRefOptions;
+  queryKey: string;
+}
+
+export function createTransportedQueryRef<TData, TVariables>(
+  options: TransportedQueryRefOptions,
+  queryKey: string,
+  promise: Promise<any>
+): InternalTransportedQueryRef<TData, TVariables> {
+  const ref: InternalTransportedQueryRef<TData, TVariables> = {
+    __transportedQueryRef: true,
+    options,
+    queryKey,
+  };
+  Object.defineProperty(ref, "toPromise", {
+    value: () => promise.then(() => ref),
+    enumerable: false,
+  });
+  return ref;
+}
+
+export function reviveTransportedQueryRef(
+  queryRef: InternalTransportedQueryRef,
+  client: ApolloClient<any>
+): [QueryRef<any, any>, CacheKey] {
+  const hydratedOptions = deserializeOptions(queryRef.options);
+  const cacheKey: CacheKey = [
+    hydratedOptions.query,
+    canonicalStringify(hydratedOptions.variables),
+    queryRef.queryKey,
+  ];
+  if (queryRef.__transportedQueryRef === true) {
+    queryRef.__transportedQueryRef = wrapQueryRef(
+      getSuspenseCache(client).getQueryRef(cacheKey, () =>
+        client.watchQuery(hydratedOptions)
+      )
+    );
+  }
+  return [queryRef.__transportedQueryRef, cacheKey];
+}
+
+function isTransportedQueryRef(
+  queryRef: object
+): queryRef is InternalTransportedQueryRef {
+  return "__transportedQueryRef" in queryRef;
+}
+
+export function useWrapTransportedQueryRef<TData, TVariables>(
+  queryRef: QueryRef<TData, TVariables> | InternalTransportedQueryRef
+): QueryRef<TData, TVariables> {
+  const client = useApolloClient();
+  let cacheKey: CacheKey | undefined;
+  let isTransported: boolean;
+  if ((isTransported = isTransportedQueryRef(queryRef))) {
+    [queryRef, cacheKey] = reviveTransportedQueryRef(queryRef, client);
+  }
+  assertWrappedQueryRef(queryRef);
+  const unwrapped = unwrapQueryRef<any>(queryRef);
+
+  useEffect(() => {
+    // We only want this to execute if the queryRef is a transported query.
+    if (!isTransported) return;
+    // We want to always keep this queryRef in the suspense cache in case another component has another instance of this transported queryRef.
+    // This effect could be removed after https://github.com/facebook/react/pull/28996 has been merged and we've updated deps to that version.
+    if (cacheKey) {
+      if (unwrapped.disposed) {
+        getSuspenseCache(client).add(cacheKey, unwrapped);
+      }
+    }
+    // Omitting the deps is intentional. This avoids stale closures and the
+    // conditional ensures we aren't running the logic on each render.
+  });
+  // Soft-retaining because useQueryRefHandlers doesn't do it for us.
+  // This effect could be removed after https://github.com/facebook/react/pull/28996 has been merged and we've updated deps to that version.
+  useEffect(() => {
+    if (isTransported) {
+      return unwrapped.softRetain();
+    }
+  }, [isTransported, unwrapped]);
```

Comment:
@jerelmiller could you please take a final look at this bit before I merge this in?